buy_together_app/tests/test_login.py

```diff
@@ -0,0 +1,95 @@
+import pytest
+from unittest.mock import Mock
+from buy_together_app import views
+from django.shortcuts import render, redirect
+from django.test import RequestFactory
+
+
```

Comment:
@pazMenachem did you decide on keeping all mocks in place eventually? On our call you talked about keeping the mock only for the `messages` call.

Also, you opted towards using the assertion methods (e.g. `assert_called_once`) while still using non-speced mock (e.g. not one which you create with create_autospec). I demonstrated on the call yesterday how this could be dangerous (e.g. if you misspell the function name, it will never fail).

Can you summarize your decisions in regards to these tests?

Reply:
since the login feature uses authenticate, which hashes the password, i didnt know how to pass an hashed password, so i stayed with the mock approach.
i forgot how it could be dangerous, eventually we are talking about typo mistake, if i use autospec, the mock knows its method, but if not, there every method i create considered legit, thou assert_called_once is built in method, im thinking maybe about the messages.info, but that mock is patched, remind me what i am missing?

Reply:
``` python
In [21]: str_mock = Mock()

In [22]: str_speced_mock = create_autospec(str, instance=True)

In [23]: str_speced_mock.capitalize.asser_not_called()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In [23], line 1
----> 1 str_speced_mock.capitalize.asser_not_called()

File /usr/lib64/python3.9/unittest/mock.py:630, in NonCallableMock.__getattr__(self, name)
    628 elif self._mock_methods is not None:
    629     if name not in self._mock_methods or name in _all_magics:
--> 630         raise AttributeError("Mock object has no attribute %r" % name)
    631 elif _is_magic(name):
    632     raise AttributeError(name)

AttributeError: Mock object has no attribute 'asser_not_called'

In [25]: str_mock.capitalize.asser_not_called()
Out[25]: <Mock name='mock.capitalize.asser_not_called()' id='140300210951600'>

In [26]: assert str_mock.capitalize.call_count == 0
```

In other words, if you misspell the assertion method then it will not cause an exception and the test will pass. 

Ultimately, best option is to use a spec - the easiest way to go about it is to use create_autospec. It will make sure your mock only mocks capabilities the mocked object actually has.

If you're not using autospec, then better not use the mock's functions (e.g. assert_called_once).

To use create_autospec, you'd need something like this:

``` python
from unittest.mock import create_autospec
@pytest.fixture
def authenticate_mock(monkeypatch):
    mock = create_autospec(views.authenticate)
    monkeypatch.setattr(views, views.authenticate.__name__, mock)
    return mock
```

Reply:
ok got it now, thanks!

Reply:
Here we create a mock object. We provide `views.authenticate` as the specifications according to which the mock should be created (what the mock should imitate):
```python
mock = create_autospec(views.authenticate)
```

Here we implant the `mock` object in-place of the `views.authenticate` function inside the `views` model:
```python
monkeypatch.setattr(views, views.authenticate.__name__, mock)
```

Reply:
got it